src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp

```diff
@@ -473,4 +522,127 @@ namespace OpenMS
     const bool compute_features { false };
     extractSpectra(experiment, targeted_exp, extracted_spectra, extracted_features, compute_features);
   }
+
+  void TargetedSpectraExtractor::matchSpectrum(
+    const MSSpectrum& input_spectrum,
+    Comparator& cmp,
+    std::vector<Match>& matches
+  )
+  {
+    // TODO: remove times debug info
+    // std::clock_t start;
+    // start = std::clock();
+    matches.clear();
+    std::vector<std::pair<Size,double>> scores;
+
+    cmp.generateScores(input_spectrum, scores, min_match_score_);
+
+    // Sort the vector of scores
+    std::sort(scores.begin(), scores.end(),
+      [](const std::pair<Size,double>& a, const std::pair<Size,double>& b)
+      {
+        return a.second > b.second;
+      });
+
+    // Set the number of best matches to return
+    const Size n = std::min(top_matches_to_report_, scores.size());
+
+    // Construct a vector of n `Match`es
+    for (Size i = 0; i < n; ++i)
+    {
+      const Size spec_idx { scores[i].first };
+      const double spec_score { scores[i].second };
+      matches.emplace_back(cmp.getLibrary()[spec_idx], spec_score);
+    }
+
+    // std::cout << "MATCH TIME: " << ((std::clock() - start) / (double)CLOCKS_PER_SEC) << std::endl;
+  }
+
+  void TargetedSpectraExtractor::targetedMatching(
+    const std::vector<MSSpectrum>& spectra,
+    Comparator& cmp,
+    FeatureMap& features
+  )
+  {
+    if (spectra.size() != features.size())
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
+    }
+
+    std::vector<Size> no_matches_idx; // to keep track of those features without a match
+    const Size tmp = top_matches_to_report_;
+    top_matches_to_report_ = 1;
+
+    for (Size i = 0; i < spectra.size(); ++i)
+    {
+      std::vector<Match> matches;
+      matchSpectrum(spectra[i], cmp, matches);
+      if (matches.size())
+      {
+        features[i].setMetaValue("spectral_library_name", matches[0].spectrum.getName());
+        features[i].setMetaValue("spectral_library_score", matches[0].score);
+        const String& comments = matches[0].spectrum.metaValueExists("Comments") ?
+          matches[0].spectrum.getMetaValue("Comments") : "";
+        features[i].setMetaValue("spectral_library_comments", comments);
+      }
+      else
+      {
+        no_matches_idx.push_back(i);
```

Comment:
Probably we should insert an empty data if no matches are found.  I have found problems when writing out the featureMap in .tsv/.csv format when metaValues differed between the features.